tests/secure_tunneling_tests.c

```diff
@@ -446,3 +628,66 @@ static int s_secure_tunneling_handle_send_data_stream_reset(struct aws_allocator
 
     return AWS_OP_SUCCESS;
 }
+
+AWS_TEST_CASE_FIXTURE(
+    secure_tunneling_handle_send_data_public,
+    before,
+    s_secure_tunneling_handle_send_data_public,
+    after,
+    &s_test_context);
+static int s_secure_tunneling_handle_send_data_public(struct aws_allocator *allocator, void *ctx) {
+    /*
+     * When a secure tunnel running in source mode sends data to destination using the public api,
+     * verify that the payload length matches what the client sends and the number of frames sent
+     * is equal to size of the payload divided by the maximum frame length.
+     */
+
+    struct secure_tunneling_test_context *test_context = ctx;
+    test_context->secure_tunnel->options->local_proxy_mode = AWS_SECURE_TUNNELING_SOURCE_MODE;
+
+    /* Open the tunnel. */
+    int rc = aws_secure_tunnel_connect(test_context->secure_tunnel);
+    ASSERT_INT_EQUALS(AWS_OP_SUCCESS, rc);
+
+    size_t buf_sizes[] = {10, 100, 1000, AWS_IOT_ST_SPLIT_MESSAGE_SIZE + 1, 2 * AWS_IOT_ST_SPLIT_MESSAGE_SIZE + 1};
+    size_t buf_sizes_len = sizeof(buf_sizes) / sizeof(buf_sizes[0]);
+
+    for (size_t i = 0; i < buf_sizes_len; ++i) {
+        /* Start a stream. */
+        s_mock_aws_websocket_send_frame_call_count = 0U;
+        s_mock_aws_websocket_send_frame_payload_len = 0U;
+        rc = aws_secure_tunnel_stream_start(test_context->secure_tunnel);
+        ASSERT_INT_EQUALS(AWS_OP_SUCCESS, rc);
+        ASSERT_UINT_EQUALS(1U, s_mock_aws_websocket_send_frame_call_count);
+        ASSERT_UINT_EQUALS(0U, s_mock_aws_websocket_send_frame_payload_len);
+
+        /* Initialize buffer of random values to send. */
+        struct aws_byte_buf buf;
+        rc = s_byte_buf_init_rand(&buf, allocator, buf_sizes[i]);
+        ASSERT_INT_EQUALS(AWS_OP_SUCCESS, rc);
+
+        struct aws_byte_cursor cur = aws_byte_cursor_from_buf(&buf);
+
+        /* Call public api to send data over secure tunnel. */
+        s_mock_aws_websocket_send_frame_call_count = 0U;
+        s_mock_aws_websocket_send_frame_payload_len = 0U;
+        rc = aws_secure_tunnel_send_data(test_context->secure_tunnel, &cur);
+        ASSERT_INT_EQUALS(AWS_OP_SUCCESS, rc);
+        int expected_call_count = (int)buf_sizes[i] / AWS_IOT_ST_SPLIT_MESSAGE_SIZE + 1;
+        ASSERT_UINT_EQUALS(expected_call_count, s_mock_aws_websocket_send_frame_call_count);
+        ASSERT_UINT_EQUALS(buf_sizes[i], s_mock_aws_websocket_send_frame_payload_len);
+
+        /* Free buffer. */
+        aws_byte_buf_clean_up(&buf);
+    }
+
+    /* Close the tunnel. */
+    aws_secure_tunnel_close(test_context->secure_tunnel);
+
+    /*
+     * FIXME(marcoaz): Is it a feature or bug that handshake_request is not released when tunnel is closed?
+     */
+    aws_http_message_release(test_context->secure_tunnel->handshake_request);
```

Comment:
I think it's safe to add a cleanup and NULLify of this to close.  At the same time, it looks like the normal place it would get cleaned up is bypassed by the mock setup and so this is test-only.

Reply:
Thanks @bretambrose I looked more closely at websocket_bootstrap.c and agree with you that `handshake_request` will be released (link below) after websocket bootstrap is complete.  Code and comment are updated.

[1] https://github.com/awslabs/aws-c-http/blob/a2a93388839add540ee26886daf4b7b9b33d0762/source/websocket_bootstrap.c#L531-L535